File: packages/validator-go/validator/validate/forbidden_scan_conformance_test.go
package validate

// Recursive forbidden-key scanning verifies the shared four-runtime fixture.
//
// tests/fixtures/spec-validity/cases.json defines SPEC §6 meta-key rejection.
// A clean specification passes; a
// forbidden meta key found at ANY depth (slot/bucket body and one level below,
// deep child subtrees, array elements, $ref-inherited bases) is a LOAD ERROR,
// never valid:true. JS / PHP / Go / Rust load this ONE file and must reproduce
// it identically. This test re-runs the real load path
// (ValidateJSON = compose → forbidden-scan → validate) against it.
//
// Each case declares `engine`: "pass" requires no load error and valid:true with
// no errors, or {code, at}
// requires a *compose.ComposeLoadError whose Code is code and whose dotted Trace
// is at. The case `files` are the composition files passed to the runtime.

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/polyspec/crudui/packages/validator-go/validator/compose"
	"github.com/polyspec/crudui/packages/validator-go/validator/internal/conformance"
)

type forbiddenScanCase struct {
	Name   string          `json:"name"`
	Note   string          `json:"note"`
	Spec   json.RawMessage `json:"spec"`
	Files  json.RawMessage `json:"files"`
	Engine json.RawMessage `json:"engine"`
}

// engineError is the {code, at} object form of a fixture engine expectation.
type engineError struct {
	Code *string `json:"code"`
	At   *string `json:"at"`
}

func loadForbiddenScanFixtures(t *testing.T) []forbiddenScanCase {
	t.Helper()
	// packages/validator-go/validator/validate → repo root is four levels up.
	path := filepath.Join("..", "..", "..", "..", "tests", "fixtures", "spec-validity", "cases.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("shared spec-validity fixture not found at %s: %v", path, err)
	}
	var cases []forbiddenScanCase
	if err := json.Unmarshal(raw, &cases); err != nil {
		t.Fatalf("fixture json decode: %v", err)
	}
	if len(cases) == 0 {
		t.Fatal("fixture is empty")
	}
	return cases
}

// filesFor decodes the optional fixture files map into the { key: rawJSON } shape
// ValidateJSON consumes.
func filesFor(t *testing.T, c forbiddenScanCase) map[string][]byte {
	t.Helper()
	files := map[string][]byte{}
	if len(c.Files) == 0 {
		return files
	}
	var fm map[string]json.RawMessage
	if err := json.Unmarshal(c.Files, &fm); err != nil {
		t.Fatalf("%s: files decode: %v", c.Name, err)
	}
	for k, v := range fm {
		files[k] = []byte(v)
	}
	return files
}

// enginePasses reports whether the fixture engine field is the literal string "pass".
func enginePasses(raw json.RawMessage) bool {
	var s string
	return json.Unmarshal(raw, &s) == nil && s == "pass"
}

func TestForbiddenScanMatchesFixture(t *testing.T) {
	for _, c := range loadForbiddenScanFixtures(t) {
		c := c
		t.Run(c.Name, func(t *testing.T) {
			conformance.Record(t, "validate", "tests/fixtures/spec-validity/cases.json", c.Name)
			// Data is irrelevant to the scan; pass empty. The scan runs in the load
			// path before any data-driven validation.
			res, err := ValidateJSON(c.Spec, nil, filesFor(t, c), "")

			if enginePasses(c.Engine) {
				if err != nil {
					t.Fatalf("clean spec must pass the load path, got error: %v", err)
				}
				if !res.Valid || len(res.Errors) != 0 {
					t.Fatalf("%s: expected valid:true with no errors, got %+v", c.Name, res)
				}
				return
			}

			var want engineError
			if uerr := json.Unmarshal(c.Engine, &want); uerr != nil || want.Code == nil || want.At == nil {
				t.Fatalf("%s: engine must be \"pass\" | {code, at}, got %s", c.Name, c.Engine)
			}
			if err == nil {
				t.Fatalf("expected load error %s at %s, but validated successfully", *want.Code, *want.At)
			}
			le, ok := err.(*compose.ComposeLoadError)
			if !ok {
				t.Fatalf("expected *compose.ComposeLoadError, got %T: %v", err, err)
			}
			if string(le.Code) != *want.Code {
				t.Fatalf("error code mismatch: expected %s, got %s (%s)", *want.Code, le.Code, le.Message)
			}
			gotPath := strings.Join(le.Trace, ".")
			if gotPath != *want.At {
				t.Fatalf("error path mismatch: expected %s, got %s", *want.At, gotPath)
			}
		})
	}
}

File: packages/validator-ts/src/forbidden-scan.conformance.test.ts
/**
 * CRUDUI recursive forbidden-scan conformance — JS reference verification.
 *
 * The shared fixture tests/fixtures/spec-validity/cases.json is the cross-
 * language contract for SPEC §6 global meta-key rejection: a clean spec
 * passes; a forbidden meta key found at ANY depth (slot/bucket body and one
 * level below, deep child subtrees, array elements, $ref-inherited bases) is a
 * LOAD ERROR, never `valid:true`. This test RE-VERIFIES the claim by running the
 * real load path (`validate` = compose → forbidden-scan → validate) against
 * the same fixture PHP / Go / Rust load.
 *
 * Each `engine: "pass"` case must return `{ valid: true, errors: [] }`. Each
 * `engine: { code, at }` case must throw a `ComposeLoadError` whose `code` is
 * `code` AND whose path (`trace`, dotted) equals `at` — the depth is
 * load-bearing, so the path is asserted, not just the code. The case `files` are
 * the composition files passed to the runtime. The meta-schema members `expect`
 * and `reason` are checked by form-metaschema.conformance.test.ts.
 *
 * Do not weaken assertions. If JS disagrees with the fixture, the fixture is NOT
 * the JS output and the cross-language contract is broken.
 */

import { describe, test, expect } from 'vitest';
import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath } from 'node:url';
import { validate, ComposeLoadError } from './validate/index';
import { provesConformance } from '../../../tests/conformance/evidence.mjs';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
// src/CRUDUI -> repo root is four levels up.
const FIXTURE = path.resolve(
  __dirname,
  '../../../tests/fixtures/spec-validity/cases.json'
);

interface FixtureCase {
  name: string;
  note: string;
  spec: Record<string, unknown>;
  files?: Record<string, Record<string, unknown>>;
  expect: 'ok' | 'fail';
  reason?: string;
  engine: 'pass' | { code: string; at: string };
}

const cases: FixtureCase[] = JSON.parse(fs.readFileSync(FIXTURE, 'utf8'));

/** Run one fixture case and record validate evidence for it. */
function proves(name: string, body: () => void): Promise<void> {
  return provesConformance(
    { features: ['validate'], fixture: 'tests/fixtures/spec-validity/cases.json', runtime: 'javascript', case: name },
    body
  );
}

function run(c: FixtureCase) {
  // Data is irrelevant to the scan; pass an empty object. The scan runs in the
  // load path before any data-driven validation.
  return validate(c.spec, {}, c.files ? { files: c.files } : {});
}

describe('forbidden-scan — clean specs pass the load path', () => {
  for (const c of cases.filter((x) => x.engine === 'pass')) {
    test(c.name, () =>
      proves(c.name, () => {
        expect(run(c)).toStrictEqual({ valid: true, errors: [] });
      }));
  }
});

describe('forbidden-scan — a forbidden meta key at any depth is a LOAD ERROR', () => {
  for (const c of cases.filter((x) => typeof x.engine === 'object')) {
    test(c.name, () =>
      proves(c.name, () => {
        const want = c.engine as { code: string; at: string };
        let thrown: unknown;
        try {
          run(c);
        } catch (e) {
          thrown = e;
        }
        expect(thrown, `${c.name} must throw a load error`).toBeInstanceOf(
          ComposeLoadError
        );
        const err = thrown as ComposeLoadError;
        expect(err.code).toStrictEqual(want.code);
        expect(err.trace.join('.')).toStrictEqual(want.at);
      }));
  }
});

describe('forbidden-scan — every fixture case declares an expectation', () => {
  test('no case is silently missing an engine field', () => {
    for (const c of cases) {
      const ok =
        c.engine === 'pass' ||
        (typeof c.engine === 'object' &&
          typeof c.engine.code === 'string' &&
          typeof c.engine.at === 'string');
      expect(ok, `${c.name} must declare engine: "pass" | {code, at}`).toBe(true);
    }
    expect(cases.length).toBeGreaterThan(0);
  });
});
